mlir/lib/Dialect/Rock/utility/loweringUtils.cpp

```diff
@@ -125,3 +128,225 @@ Type mlir::rock::vectorTypeOrSelf(Type elementType, int64_t len) {
+    LowerDimPartInfo kDimIterPartInfo, LowerDimPartInfo dDimIterPartInfo) {
+  if (dName != "m" && dName != "n") {
+    return emitError(loc, "expected dName to be m or n but got " + dName);
+  }
+  StringRef thisBlockDim = dName == "m" ? "m_block" : "n_block";
+  StringRef otherBlockDim = dName == "m" ? "n_block" : "m_block";
+
+  MemRefType matrixType = globalBuffer.getType().cast<MemRefType>();
+  ArrayRef<int64_t> matrixShape = matrixType.getShape();
+  int64_t kGlobal = matrixShape[1];
+  int64_t dGlobal = matrixShape[2];
+
+  int64_t kIters = kGlobal / kPerBlock;
+  int64_t dataPerThread = (kPerBlock * dPerBlock) / blockSize;
+
+  // Note: (kThreads * dThreads) = (kPerBlock * dPerBlock) / dataPerThread) =
+  // blockSize
+  GPUViews gpuViews;
+  {
+    TopDownTMBuilder gridwiseSplitId(
+        b, {"k_loop", "g_block", "m_block", "n_block", "tid", "iter"},
+        {kIters, bidGridLengths[0], bidGridLengths[1], bidGridLengths[2],
+         blockSize, dataPerThread},
+        loc);
+    gridwiseSplitId.passThrough({"k_loop", "g_block", "m_block", "n_block"});
+    auto [tidDimNames, tidDimSizes] =
+        getCombinedOrder(kDimTidPartInfo, dDimTidPartInfo);
+    gridwiseSplitId.merge(tidDimNames, getOutDimIndices(4, tidDimNames.size()),
+                          "tid", tidDimSizes);
+    auto [iterDimNames, iterDimSizes] =
+        getCombinedOrder(kDimIterPartInfo, dDimIterPartInfo);
+    gridwiseSplitId.merge(
+        iterDimNames,
+        getOutDimIndices(4 + tidDimNames.size(), iterDimNames.size()), "iter",
+        iterDimSizes);
+
+    TransformMapAttr splitIdAttr = gridwiseSplitId.get();
+    auto toGlobalIdx = TopDownTMBuilder::below(gridwiseSplitId, splitIdAttr);
+    toGlobalIdx.passThrough({"g"}, {0}, {"g_block"});
+    {
+      SmallVector<StringRef, 3> kSubDimNames{"k_loop"};
+      kSubDimNames.append(kDimTidPartInfo.lowerDimPartNames.begin(),
+                          kDimTidPartInfo.lowerDimPartNames.end());
+      kSubDimNames.append(kDimIterPartInfo.lowerDimPartNames.begin(),
+                          kDimIterPartInfo.lowerDimPartNames.end());
+      SmallVector<int64_t, 3> kSubDimSizes{kIters};
+      kSubDimSizes.append(kDimTidPartInfo.lowerDimPartSizes.begin(),
+                          kDimTidPartInfo.lowerDimPartSizes.end());
+      kSubDimSizes.append(kDimIterPartInfo.lowerDimPartSizes.begin(),
+                          kDimIterPartInfo.lowerDimPartSizes.end());
+      toGlobalIdx.unmerge("k", 1, kSubDimNames, kSubDimSizes);
+    }
+
+    {
+      SmallVector<StringRef, 3> dSubDimNames{thisBlockDim};
+      dSubDimNames.append(dDimTidPartInfo.lowerDimPartNames.begin(),
+                          dDimTidPartInfo.lowerDimPartNames.end());
+      dSubDimNames.append(dDimIterPartInfo.lowerDimPartNames.begin(),
+                          dDimIterPartInfo.lowerDimPartNames.end());
+      SmallVector<int64_t, 3> dSubDimSizes{dGlobal / dPerBlock};
+      dSubDimSizes.append(dDimTidPartInfo.lowerDimPartSizes.begin(),
```

Comment:
I'm a bit suspicious of this

This looks like uniformly assuming that the iter dimension moves faster than the tid dimension, which doesn't feel right.

Reply:
it works between :

1) loaded tile from global
2) reshuffled tile from global

and yea it wont work for others -- not as general that bit.
I could just simply make two independent view creators.